File: app-android/app/src/main/java/com/goexercise/app/domain/CatRank.kt
package com.goexercise.app.domain

import com.goexercise.app.domain.friends.FriendProfile

/**
 * 称号メタル種別。iOS `MetalKind` の移植。色合い/演出の段階を表す。
 */
enum class MetalKind {
    Bronze,
    BronzePlus,
    Silver,
    SilverPlus,
    GoldMinus,
    Gold,
    GoldPlus,
    Platinum,
    Rainbow,
}

/**
 * 連続記録ベースの 11 段階ランク。iOS `CatRank` の純ロジック移植。
 *
 * rank = 連続日数 `streak` に対し `thresholds` のうち `<= max(0, streak)` を満たす個数(0..11)。
 * 称号/メタルは rank に従って決まり、rank0 は称号なし(null)。
 * iconSymbol(SF Symbol)は iOS 専用のため Android では持たない(Compose アイコンは別途)。
 *
 * 純 Kotlin のみ(Android/Compose import なし)で JVM ユニットテスト可能。
 */
class CatRank private constructor(val rank: Int) {

    /** 0.0..1.0 の充実度(rank/11)。背景進化や演出の強度に使う。 */
    val richness: Double
        get() = rank.toDouble() / 11.0

    /** 称号名。rank0 は null。 */
    val title: String?
        get() = when (rank) {
            1 -> "みならいネコ"
            2 -> "かけだしネコ"
            3 -> "がんばりネコ"
            4 -> "まいにちネコ"
            5 -> "きたえネコ"
            6 -> "つわものネコ"
            7 -> "ベテランネコ"
            8 -> "達人ネコ"
            9 -> "仙人ネコ"
            10 -> "レジェンドネコ"
            11 -> "ぬしネコ"
            else -> null
        }

    /** 称号メタル種別。rank0 は null。 */
    val metalKind: MetalKind?
        get() = when (rank) {
            1 -> MetalKind.Bronze
            2 -> MetalKind.BronzePlus
            3 -> MetalKind.Silver
            4 -> MetalKind.SilverPlus
            5 -> MetalKind.GoldMinus
            6, 7 -> MetalKind.Gold
            8, 9 -> MetalKind.GoldPlus
            10 -> MetalKind.Platinum
            11 -> MetalKind.Rainbow
            else -> null
        }

    override fun equals(other: Any?): Boolean = other is CatRank && other.rank == rank
    override fun hashCode(): Int = rank
    override fun toString(): String = "CatRank(rank=$rank, title=$title)"

    companion object {
        /** 各ランクへ昇格する連続日数の閾値(11 段階)。 */
        val thresholds = listOf(7, 14, 30, 50, 75, 100, 150, 200, 300, 365, 500)

        /** 連続日数からランクを生成。負値は 0 に丸める。 */
        fun of(currentStreak: Int): CatRank {
            val streak = maxOf(0, currentStreak)
            val rank = thresholds.count { it <= streak }
            return CatRank(rank)
        }
    }
}

/** 友達(または自分)の現在連続記録から導かれる称号ランク。 */
val FriendProfile.rank: CatRank
    get() = CatRank.of(currentStreak)

File: app/GOExercise/GOExercise/Models/CatRank.swift
import Foundation

/// メタルの種別。色(SwiftUI)は `MetalStyle.swift` が担う。
/// `+`/`-` は同系統 ±8% 明度のバリアント。`rainbow` は AngularGradient。
enum MetalKind: Equatable, Sendable {
    case bronze, bronzePlus
    case silver, silverPlus
    case goldMinus, gold, goldPlus
    case platinum
    case rainbow
}

/// 達成の豪華演出の背骨。**現在の連続日数**で決まる統一11段ランク。
/// 連続が途切れたら rank も落ちる(降格)= 守る動機。フリーズで連続が保たれれば rank も保たれる。
/// A背景の濃さ・C称号名・Cメタル色・B昇格判定の唯一の入力。
struct CatRank: Equatable, Sendable {
    /// 連続閾値(昇順)。rank = この配列のうち streak 以下の要素数。
    static let thresholds = [7, 14, 30, 50, 75, 100, 150, 200, 300, 365, 500]

    /// 0..11。0 = 称号なし(連続7日未満)。
    let rank: Int

    init(currentStreak: Int) {
        let s = max(0, currentStreak)
        self.rank = Self.thresholds.filter { s >= $0 }.count
    }

    /// 0..1。背景 richness 等の連続的な濃さ。
    var richness: Double { Double(rank) / Double(Self.thresholds.count) }

    /// 称号(案B「ねこ仕立て」)。rank0 は nil。
    var title: String? {
        switch rank {
        case 1: return "みならいネコ"
        case 2: return "かけだしネコ"
        case 3: return "がんばりネコ"
        case 4: return "まいにちネコ"
        case 5: return "きたえネコ"
        case 6: return "つわものネコ"
        case 7: return "ベテランネコ"
        case 8: return "達人ネコ"
        case 9: return "仙人ネコ"
        case 10: return "レジェンドネコ"
        case 11: return "ぬしネコ"
        default: return nil
        }
    }

    /// メタル種別。rank0 は nil。
    var metalKind: MetalKind? {
        switch rank {
        case 1: return .bronze
        case 2: return .bronzePlus
        case 3: return .silver
        case 4: return .silverPlus
        case 5: return .goldMinus
        case 6, 7: return .gold
        case 8, 9: return .goldPlus
        case 10: return .platinum
        case 11: return .rainbow
        default: return nil
        }
    }

    /// 称号バッジの小アイコン(SF Symbol)。ブランド一貫性のため全 rank 共通の肉球。
    var iconSymbol: String { "pawprint.fill" }
}
